lib/model.js

```diff
@@ -1438,29 +1440,82 @@ Model.syncIndexes = function syncIndexes(options, callback) {
           }
           cb(null, dropped);
         });
+      }); */
+
+      this.diffIndexes((err, result) => {
+        console.log('the result', result.toDrop);
+        if (err != null) {
+          return cb(err);
+        }
+        this.wipeIndexes(result.toDrop, err => {
+          if (err != null) {
+            return cb(err);
+          }
+          this.createIndexes(options, err => {
+            if (err != null) {
+              return cb(err);
+            }
+            cb(null, result.toDrop);
+          });
+        });
       });
     });
   }, this.events);
 };
 
+
+Model.wipeIndexes = function wipeIndexes(destroy, callback) {
+  console.log('wipe Indexes', destroy);
+  _checkContext(this, 'wipeIndexes');
+
+  callback = this.$handleCallbackError(callback);
+
+  return this.db.base._promiseOrCallback(callback, cb => {
+    const collection = this.$__collection;
+
+    if (destroy.length === 0) {
+      return cb(null, []);
+    }
+
+    dropIndexes(destroy, cb);
+
+    function dropIndexes(toDrop, cb) {
+      let remaining = toDrop.length;
+      let error = false;
+      toDrop.forEach(indexName => {
+        collection.dropIndex(indexName, err => {
+          if (err != null) {
+            error = true;
+            return cb(err);
+          }
+          if (!error) {
+            --remaining || cb(null, toDrop);
+          }
+        });
+      });
+    }
+  });
+};
+
+
 /**
  * Does a dry-run of Model.syncIndexes(), meaning that
  * the result of this function would be the result of
  * Model.syncIndexes().
  *
- * @param {Object} options not used at all.
  * @param {Function} callback optional callback
- * @returns {Promise} which containts an object, {toDrop, toCreate}, which
+ * @returns {Promise} which contains an object, {toDrop, toCreate}, which
  * are indexes that would be dropped in mongodb and indexes that would be created in mongodb.
  */
 
-Model.diffIndexes = function diffIndexes(options, callback) {
+Model.diffIndexes = function diffIndexes(callback) {
   const toDrop = [];
   const toCreate = [];
   callback = this.$handleCallbackError(callback);
   return this.db.base._promiseOrCallback(callback, cb => {
     cb = this.$wrapCallback(cb);
     this.listIndexes((err, indexes) => {
```

Comment:
We're missing some logic here from `cleanIndexes()` that removes non-related indexes: https://github.com/Automattic/mongoose/blob/f5fe027d41d994bb6b11468495aae80232fcb740/lib/model.js#L1536-L1537 .

We should take the logic that determines which indexes to drop, https://github.com/Automattic/mongoose/blob/f5fe027d41d994bb6b11468495aae80232fcb740/lib/model.js#L1536-L1559, and put it behind a separate helper function `function indexesToDrop(dbIndexes, schemaIndexes) { ... }`. `diffIndexes()` and `cleanIndexes()` should both call `indexesToDrop()`.

Reply:
Should I do the same with create?